### DISCUSSION TITLE
Fitting Second-Order Systems (e.g., Duffing Oscillator) – How to Handle First-Order Variables?

### DISCUSSION THREAD

**Discussion by abelleipardo** [dynamicslab/pysindy#644] [General]
I would like to fit a second-order dynamical system, specifically something like the Duffing oscillator. When written in first-order form, the full system is:
xdot = X
ydot = Y
Xdot = fcn1(x,y)
Ydot = fcn2(x,y)

In the paper "_A Unified Sparse Optimization Framework to Learn Parsimonious Physics-Informed Models From Data_" by Champion et al. (2020), the authors state:
“Because the variables X and Y are fixed to represent momentum, we apply SINDy only to fit the second two equations.”

I would like some clarification on how to implement this in practice. Specifically:
- Should I manually compute xdot and ydot from my simulation data, i.e., treat X and Y as known (from the simulation), and then only fit Xdot and Ydot using SINDy?
- If so, how should I prepare the data for SINDy so it understands which variables to treat as known and which to fit?
- I’m using the ConstrainedSR3 optimizer—are there any special considerations when applying it in this context?

Thank you for your help!

**Comment by Jacob-Stevens-Haas** [comment]
Hey, thanks for your question! 
* You have options, but yes, you need to manually compute X and Y (potentially using `pysindy.differentiation`).  You _could_ fit all four variables (`x, X, y, Y`) and let SINDy discover the first two equations, but to do as Champion et al, you need to only fit Xdot and Ydot
* To do this, you would pass in `X` and `Y` as your input variable. Then you would pass in `x` and `y` as control variables.  To wit:
```python
model.fit(momentum, u=position)
```
where `momentum` is an array containing `X` and `Y` stacked along the last axis and `position` comprises `x` and `y` stacked likewise. 
* Nope, not that I can think of.  Two things to know with writing constraints:
1. you need to know the function library ordering, so it helps to use `feature_lib.fit(arr).get_feature_names()` before writing constraints.
2. Constraint format has changed between 1.7.5 and 2.0.0, so check the documentation of whatever version is installed.  It may not line up with how the examples create constraints.

**Reply by abelleipardo** [reply]
Thank you for your response, Jacob. 
I have been working on my code with your approach. In addition to my previous input, I would like to incorporate parameterized dynamics. To this end, I am following the logic from the Logistic map and Hopf normal form from https://pysindy.readthedocs.io/en/latest/examples/sindy-original-example/30f4dcd/original_paper.html by introducing a third equation mudot = 0.
I have the following questions:
1. Would you recommend I add this new variable in the momentum equation and let SINDy discover mudot = 0, or should I add it to the known control variables?

2. Are there other ways to incorporate parameterized dynamics?

3. [unrelated to parameterized dynamics] Following your suggestion about looking at the library ordering, I am running into an issue with building a custom library. I create a custom library in the following way:

nonlinear_library_funcs = [
    lambda q, dq: q**2 * dq]
nonlinear_library_names = [
    lambda q, dq: 'q^2*dq']
nonlinear_library = ps.CustomLibrary(
    library_functions=nonlinear_library_funcs,
    function_names=nonlinear_library_names
)
sample_momentum = momentum_data_list[0]
sample_position = position_data_list[0]
sample_input_arr = np.hstack((sample_momentum, sample_position))
feature_names = nonlinear_library.fit(sample_input_arr).get_feature_names()

I then get feature_names = ['q^2*dq', 'q^2*dq', 'q^2*dq', ...] repeated 36 times.
Why is this happening and how should I fix it?
A work around I found (by asking Gemini how to fix it) was to write this class before creating the library:

class PredefinedLibrary(ps.CustomLibrary):
    def get_feature_names(self, input_features=None):
        # Manually create the names from the function signatures
        names = []
        for func in self.function_names:
            # Get the argument names from the function's code object
            arg_names = func.__code__.co_varnames[:func.__code__.co_argcount]
            # Call the naming function with dummy string args
            names.append(func(*arg_names))
        return names
 
 However, I suspect I am making a mistake when defining the library.

Thank you for your help!

Andrés

**Reply by Jacob-Stevens-Haas** [reply]
(I've added \`\`\` to format your code)

> should I add it to the known control variables?

Add it to the control variables.  The original notebook predates the introduction of controls to pysindy.  Lines like:
```python
x_train[ic_idx][:, 2] = mu
```
Happen under-the-hood now

> Are there other ways to incorporate parameterized dynamics?

It's somewhat common to do parameterized dynamics with a `GeneralizedLibrary`.  This allows you to choose one library for functions of the state variables, a second one for parameters, and multiply them.  For `WeakSINDy`, you'd want the `ParameterizedLibrary`, which is like `GeneralizedLibrary` but a bit of a hack because the under-the-hood integration by parts behaves differently for multiplication.

> I am running into an issue with building a custom library
```python
lambda q, dq: 'q^2*dq'
```
is a function that takes two inputs and returns a constant string.  On the other hand,
```python
lambda q, dq: f'{q}^2*{dq}'
```
is a function that takes two inputs and returns a template string with the inputs inserted in position.

**Reply by abelleipardo** [reply]
Thank you again, Jacob.
I'm working on discovering a parameterized model for a fluid-structure interaction (FSI) problem. I am using high-fidelity CFD / Structural data.
I implemented your suggested approach and have my parameterized variable in the control variables using Constrained SR3. My library is a `GeneralizedLibrary `, combining a `PolynomialLibrary` (degree 1 and 2) with a `CustomLibrary` for my parameter (e.g., `mu`, `exp(-mu)`, `1/mu`).

I am running into some challenges and would appreciate some guidance on how to move forward:
- For weak regularization (`reg_weights_lam < 0.005`), the optimizer fails to converge after tens of thousand of iterations with the error decreasing very slowly. Stronger regularization converges but models become physically trivial, zeroing out some equations.
- All discovered models are unstable when simulated forward in time.

**Specific questions:**
1. I have only used `relax_coeff_nu = 1.0`, while iterating over `reg_weights_lam`. I don't have a predefined target for the threshold `eta`. Would you suggest changing the value of the `relax_coeff_nu`? How? `nu = 1/lam`?
2. I would like to implement the unbiasing procedure at the end of the fitting, as described by Champion et al. (2020): "after the respective optimization algorithm is applied to select the correct terms in the dynamics, a least squares fit is performed on just these terms to find the coefficient values". To re-fit the coefficients on the discovered model while preserving the constraints, what is the correct implementation? My guess is to run ConstrainedSR3 on the sparse library and set `reg_weights_lam = 0`?
3. In addition, I implemented Trapping SINDy fitting all equations, because I could not use the approach `model.fit(momentum, u=position)`. Trapping SINDy is memory-intensive, and I was also not able to reach convergence. I am using `reg_weights_lam = 0` as a first approach, in order to avoid using cvxpy which is very slow.
3.1. Is there any special consideration I have to take into account to use the momentum approach?
3.2. There are many user-defined hyper-parameters in Trapping SINDy, and the order of magnitude of some of them changes in different examples (https://github.com/dynamicslab/pysindy/tree/benchmarking/examples/8_trapping_sindy_examples), do you have any recommendation on how to start tuning them?
3.3. Only polynomial terms of degree 2 are admitted in this library, right? Is there any way I can try to include the exponential term for example?
5. Do you have any other suggestions?

Thank you so much for your time!

Best,
Andrés
